### PR TITLE
Fix customer deletion and payment detachment in staging site

### LINKED ISSUE #2614: Deleting customer on staging site detaches tokens from customer at Stripe, breaking live site saved payment methods (and subscriptions)
**Describe the bug**
I've been told that if an admin user deletes WordPress users on a staging site (or any other duplicate copy of the live site, like a local dev copy), payment method tokens will be detached from customers on Stripe's side. In the case of a store with WC Subscriptions running, this can break future renewal payments.

**To Reproduce**
1. Go to `wp-admin` dashboard for a site with Stripe plugin
1. Go to **Users > All Users**
1. Click on radio box next to a user that has a saved payment method with Stripe
1. Click on the **Bulk actions** select field
1. Click on **Delete** option
1. Click **Apply**
1. Go to Stripe.com and open the dashboard for the corresponding Stripe account
1. Confirm customer is missing payment tokens 

**Expected behavior**
The payment method deletion or disconnection is expected behaviour when users are deleted form a live site; however, a staging site's activities should not interupt the activity on the live site. Therefore Stripe.com data, like customers and payment methods, should not be deleted when deleting corresponding data from a staging site.

This should be the case regardless of whether WC Subscriptions is running or not, even though the expected behaviour is most immediately impacted when running WC Subscriptions.

**Additional context**

I haven't reproduced this yet, but wanted to open an issue for this to be investigated after @NickGreen mentioned the behaviour to me yesterday.

### PR DESCRIPTION
**Pull request by Mayisha**
Fixes #2614 

When a user is deleted from a staging site, payment method tokens are getting detached from customers on Stripe's side. This hampers the customer's account and subscriptions on the production site. The expected behavior is not to detach the payment method from the Stripe customer if the user is deleted from a staging site. If the payment method deletion takes place by customer interaction (from `My account` page of the customer), then the method should be detached for staging or production any site.

## Changes proposed in this Pull Request:
- Added `should_detach_payment_method_from_customer` function to check if we should make the detach payment method request to Stripe.
- This function returns false if the site is a staging site and the request is coming from the admin dashboard (which is mainly the delete user request). Used WP method `wp_get_environment_type` to determine the environment.
- This function returns true for all other cases and allows sending a detach payment method request to Stripe.

## Testing instructions
- For the sake of testing in the local environment, you may replace `staging` with `development` in the condition check [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672/files#diff-9b5ad4eb8d7bffb53e706353d9fbc116c6d4f6c8f328eb154510ed24fe351b5aR370).
- Purchase a subscription with a new user `U`.
- With this new user go to the **My account -> Payment methods** page and add 3 more payment methods there.
- Find the corresponding customer in your Stripe dashboard.
- Ensure that the customer has a total of 4 payment methods attached.
- Now delete a payment method from **My account -> Payment methods** page (except the one used for subscription)
- Check the customer in your Stripe dashboard again and ensure that it has 3 payment methods attached now. As the payment method was deleted directly by the customer's interaction, we should detach the payment method here.
- Now go to **Users > All Users** in the `wp-admin` as an admin user.
- Find user `U` and delete the user.
- Check the customer in your Stripe dashboard again and ensure that it still has the 3 payment methods attached.

### CODE REVIEW

File: includes/class-wc-stripe-api.php [hunk 1]
```diff
@@ -341,4 +345,32 @@ public static function detach_payment_method_from_customer( string $customer_id,
...
+
+		// If we are in test mode, we can always detach the payment method.
+		if ( $test_mode ) {
+			return true;
+		}
+
+		// Return false for the delete user request from the admin dashboard when the site is a staging site.
+		// This is to avoid detaching the payment method from the production site.
+		// Requests coming from the customer account page i.e delete payment method, are not affected by this and returns true.
+		if ( 'staging' === wp_get_environment_type() && is_admin() ) {
```
Comment (Mayisha) [inline includes/class-wc-stripe-api.php]: This does not protect from detaching a payment method from the local dev copy of the production site though. Alternatively, we can use the following condition-

```
if ( is_admin() ) {
			if ( 'production' === wp_get_environment_type() ) {
				return true;
			} else {
				return false;
			}
		}
```

This however will keep the payment method attached to the customer when the delete user request is from a local dev environment which is in Stripe live mode, irrespective of the site being a copy of a production site or a completely local dev site.

Comment (james-allan) [reply includes/class-wc-stripe-api.php]: I've done some digging around and it seems like some plugins use `wp_get_environment_type()` somewhat consistently.

WC core uses the Jetpack::Status class which has a `is_staging_site()` function. That function has fallback mechanisms which use a list of known url regex patterns essentially to check if the url includes some form of `.staging.com` for example. 

However when it uses `wp_get_environment_type()` it considers "staging" to be any type that is not `'production'` or `'local'`. Which by process of elimination means `staging = 'development' or 'staging'`.

WooPayments also uses `wp_get_environment_type()` as well in [`WCPay\Core\Mode`](https://github.com/Automattic/woocommerce-payments/blob/6.1.0/includes/core/class-mode.php#L86). In that class it also considers dev sites to be `development` or `staging` environments. 

I don't think we need to copy them per se, I guess it just depends on what we want to do in each environment. 

IMO I'd think we'd only want to detach payment methods on production sites. Having said that, I don't know if I can distinguish between `development`, `local` or `staging` in this scenario. It sounds like you might have an idea on that though with what you said here:

> This however will keep the payment method attached to the customer when the delete user request is from a local dev environment which is in Stripe live mode, irrespective of the site being a copy of a production site or a completely local dev site.
 
To me all three `development`, `local` and `staging` indicate a duplicate copy of the site for testing purposes and a site we wouldn't want to send live destructive requests from. From your comment above, it sounds like you're leaning towards local dev sites legitimately detaching payment methods from customers. Is that right?

Comment (Mayisha) [reply includes/class-wc-stripe-api.php]: Thanks @james-allan for the references. 

> From your comment above, it sounds like you're leaning towards local dev sites legitimately detaching payment methods from customers. Is that right?

I initially made the comment to seek the reviewer's opinion as I was unsure which one was the best solution. 
After going through the references you mentioned, I am leaning towards detaching the method from the production site only as in other places also staging and development sites are considered the same.

> To me all three development, local and staging indicate a duplicate copy of the site for testing purposes and a site we wouldn't want to send live destructive requests from.

This is also a good point.

I have made the changes in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672/commits/5e5aae8d89b0496393ddab268e76794fe05469db accordingly.

**Review by james-allan — APPROVED** [review]
Looks good to me 👍 

I just pushed a commit (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2672/commits/74b0335c971401eb5d3721b9eecd52115a76880b) to fix up the function block comment. There were param tags most likely copied from the function above. :) 

I tested the following scenarios: 

1. Deleting a payment method from the customer's my account page.
   - [x] In test mode it was deleted from Stripe 👍 
   - [x] Not in test mode, it was also deleted from Stripe 👍
   - **All deletes from the front end, irrespective of the environment, do result in a delete in Stripe.**
1. Deleting a customer from the admin.
   - [x] In test mode it was deleted from Stripe 👍 
   - [x] In live mode on a "local" site it wasn't deleted. 👍 
   - [x] In live mode on a "production" site it was deleted. 👍

**Pull request merged**